### PR TITLE
Add Protobuf support for trino-kafka

### PR DESCRIPTION
**Pull request by nevillelyh**
## Description

Add Protobuf support for Kafka

This is a contribution from the [Starburst Kafka Connector](https://docs.starburst.io/latest/connector/starburst-kafka.html)

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add Protobuf message encoding/decoding support for Kafka connector
```

### CODE REVIEW

**Comment by nevillelyh** [comment]
Code, tests & docs are ready. Still missing product tests, will add later.

File: pom.xml [hunk 1]
```diff
@@ -1362,6 +1383,20 @@
                 </exclusions>
             </dependency>
 
+            <dependency>
+                <groupId>io.confluent</groupId>
+                <artifactId>kafka-protobuf-provider</artifactId>
+                <version>${dep.confluent.version}</version>
```
Comment (nevillelyh) [inline pom.xml:1389]: This should also be `provided` due to license but it's needed in `TestKafkaProtobufReads` product tests.
https://github.com/trinodb/trino/pull/14734/files#diff-7cdd855d210d35e291a669e4d8bc34963e693ad07501737f5b0fc317f1ba3c67R25

**Comment by nevillelyh** [comment]
It's still failing `suite-6-non-generic` after I fixed the `kafka-protobuf-provider` dependency in test containers. Cannot reproduce locally. Possibly a caching issue?

```
presto-master       | java.lang.NoClassDefFoundError: io/confluent/kafka/schemaregistry/protobuf/ProtobufSchemaProvider
presto-master       | 	at io.trino.plugin.kafka.schema.confluent.ConfluentModule$LazyLoadedProtobufSchemaProvider.create(ConfluentModule.java:186)
presto-master       | 	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:183)
presto-master       | 	at io.trino.plugin.kafka.schema.confluent.ConfluentModule$LazyLoadedProtobufSchemaProvider.parseSchema(ConfluentModule.java:181)
presto-master       | 	at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema(CachedSchemaRegistryClient.java:197)
presto-master       | 	at io.trino.plugin.kafka.schema.confluent.ClassLoaderSafeSchemaRegistryClient.parseSchema(ClassLoaderSafeSchemaRegistryClient.java:49)
```

**Comment by michaelpearce-gain** [comment]
One thing looking at this i couldnt spot where the smarts is to auto deduce if to use proto or avro when using schema registry client, where a broker may have topics with both

Schema registry holds the schematype now and like wise supports avro, and proto, so just from the schema id or from the subject latest schema you can auto deduce if its avro or proto, thus easily allowing to switch based on this which deserializing method to use (avro vs proto)

    SchemaRegistryClient src = <init>

    ParsedSchema parsedSchema = src.getSchemaById(0);
    parsedSchema.schemaType()

**Comment by nevillelyh** [comment]
Tests finally green. @Praveen2112 PTAL when you get a chance? 🙏

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/protobuf/ProtobufUtils.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                        .addField(
+                                processType(
+                                        FieldDescriptorProto.newBuilder()
+                                                .setName("value")
+                                                .setNumber(2)
+                                                .setLabel(Label.LABEL_OPTIONAL),
+                                        protoType.getValueType(),
+                                        definedMessages))
+                                .build());
+                // Handle for underscores and name
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/protobuf/ProtobufUtils.java:174]: What do we want to test here, that the descriptor is constructed correctly? That's covered by `TestProtobuf{Encoder,Decoder}` already. Commenting out anything except `.setName(field.getName())` fails the tests.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConnectorModule.java [hunk 3]
```diff
@@ -58,6 +59,7 @@ public void setup(Binder binder)
 
         configBinder(binder).bindConfig(KafkaConfig.class);
         bindTopicSchemaProviderModule(FileTableDescriptionSupplier.NAME, new FileTableDescriptionSupplierModule());
+        bindTopicSchemaProviderModule(FileTableDescriptionSupplier.NAME, new ProtobufModule());
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConnectorModule.java]: Done. Kept them as 2 separate commits but can squash once we're happy with the changes.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java [hunk 4]
```diff
@@ -11,7 +11,21 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package io.trino.plugin.kafka.protobuf;
+
```
Comment (Praveen2112) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:14]: License has to be removed.

**Review by Praveen2112 — COMMENTED** [review]
I think we could have a single commit where we could implement the `Protobuf` in the decoder module and use them in Kafka connector for now.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java (hunk 4 above)
Comment (nevillelyh) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java]: Duh IntelliJ 🤦
Fixed.

File: lib/trino-record-decoder/pom.xml [hunk 5]
```diff
@@ -43,6 +49,16 @@
...
+            <groupId>com.google.protobuf</groupId>
+            <artifactId>protobuf-java</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.squareup.wire</groupId>
+            <artifactId>wire-schema</artifactId>
+        </dependency>
+
         <dependency>
```
Comment (Praveen2112) [inline lib/trino-record-decoder/pom.xml:62]: Should we exclude few resources for duplicate classfinder to be happy ?

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/FixedSchemaDynamicMessageProvider.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    public static class Factory
+            implements DynamicMessageProvider.Factory
+    {
+        @Override
+        public DynamicMessageProvider create(Optional<String> protoFile)
+        {
+            checkState(protoFile.isPresent(), "proto file is missing");
+            try {
+                Descriptor descriptor = ProtobufUtils.getFileDescriptor(protoFile.get()).findMessageTypeByName(DEFAULT_MESSAGE);
```
Comment (Praveen2112) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/FixedSchemaDynamicMessageProvider.java:58]: Can we use `protoFile#orElseThrow` as `Optional#get` will be deprecated

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            Descriptors.FieldDescriptor fieldDescriptor = valueDescriptor.get().findFieldByName(pathElement);
+            value = ((DynamicMessage) value).getField(fieldDescriptor);
+            valueDescriptor = getDescriptor(fieldDescriptor);
+        }
+        return value;
+    }
+
+    private static Optional<Descriptors.Descriptor> getDescriptor(Descriptors.FieldDescriptor fieldDescriptor)
+    {
+        if (fieldDescriptor.getJavaType() == Descriptors.FieldDescriptor.JavaType.MESSAGE) {
```
Comment (Praveen2112) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:141]: Can we use static import for `MESSAGE`

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                        ProtobufColumnDecoder::new));
+    }
+
+    @Override
+    public Optional<Map<DecoderColumnHandle, FieldValueProvider>> decodeRow(byte[] data)
+    {
+        return Optional.of(columnDecoders.entrySet().stream()
+                .collect(toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> entry.getValue().decodeField(dynamicMessageProvider.parseDynamicMessage(data)))));
```
Comment (Praveen2112) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java:51]: Can we parse it once and use the `DynamicMessage` for all the required columns.

**Review by Praveen2112 — APPROVED** [review]
LGTM Minor comments

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java (hunk 8 above)
Comment (nevillelyh) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java:51]: Good catch!

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufRowEncoder.java [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ProtobufRowEncoder
+        extends AbstractRowEncoder
+{
+    public static final String NAME = "protobuf";
+
+    private static final Set<Type> SUPPORTED_PRIMITIVE_TYPES = ImmutableSet.of(
+            BOOLEAN, INTEGER, BIGINT, DOUBLE, REAL);
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufRowEncoder.java:69]: https://github.com/trinodb/trino/issues/15012

File: docs/src/main/sphinx/connector/kafka.rst [hunk 10]
```diff
@@ -1237,6 +1348,76 @@ The schema evolution behavior is as follows:
...
+
+===================================== =======================================
+Trino data type                       Allowed Protobuf data type
+===================================== =======================================
+``BOOLEAN``                           ``bool``
+``INTEGER``                           ``int32``, ``uint32``, ``sint32``, ``fixed32``, ``sfixed32``
+``BIGINT``                            ``int64``, ``uint64``, ``sint64``, ``fixed64``, ``sfixed64``
+``DOUBLE``                            ``double``
+``REAL``                              ``float``
+``VARCHAR`` / ``VARCHAR(x)``          ``string``
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply docs/src/main/sphinx/connector/kafka.rst:1383]: Seems that enum maps to `varchar` for both encoder & decoder, added both.

File: lib/trino-record-decoder/pom.xml (hunk 5 above)
Comment (nevillelyh) [reply lib/trino-record-decoder/pom.xml:62]: Looks like Guava is the only one we can exclude? Kotlin is already pinned to make enforcer happy.
https://mvnrepository.com/artifact/com.squareup.wire/wire-schema/3.2.2

Speaking of, do we want a more recent version of this? They removed Guava dependency but bumped Kotlin, which should be backwards compatible.
https://mvnrepository.com/artifact/com.squareup.wire/wire-schema/4.4.3

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    @Nullable
+    private static Object locateField(DynamicMessage message, String columnMapping)
+    {
+        Object value = message;
+        Optional<Descriptors.Descriptor> valueDescriptor = Optional.of(message.getDescriptorForType());
+        for (String pathElement : Splitter.on('/').omitEmptyStrings().split(columnMapping)) {
+            if (valueDescriptor.filter(descriptor -> descriptor.findFieldByName(pathElement) != null).isEmpty()) {
+                return null;
+            }
+            Descriptors.FieldDescriptor fieldDescriptor = valueDescriptor.get().findFieldByName(pathElement);
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java]: Also fixed a few more `Descriptors.*` imports

**Comment by Praveen2112** [comment]
Let me know once the changes are applied.

**Comment by nevillelyh** [comment]
@Praveen2112 hey yeah I addressed most except the question regarding excisions https://github.com/trinodb/trino/pull/14734#discussion_r1021639358
Thoughts?

**Review by Praveen2112 — COMMENTED** [review]
Some of the changes were not applied.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufValueProvider.java [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        else {
+            blockBuilder = type.createBlockBuilder(null, 1);
+        }
+        BlockBuilder singleRowBuilder = blockBuilder.beginBlockEntry();
+        DynamicMessage record = (DynamicMessage) value;
+        List<RowType.Field> fields = ((RowType) type).getFields();
+        for (RowType.Field field : fields) {
+            checkState(field.getName().isPresent(), "field name not found");
+            Descriptors.FieldDescriptor fieldDescriptor = getFieldDescriptor(record, field.getName().get());
+            checkState(fieldDescriptor != null, format("Unknown Field %s", field.getName().get()));
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufValueProvider.java:288]: Uh missed it. Fixed now.

File: pom.xml [hunk 13]
```diff
@@ -2014,6 +2048,9 @@
                                 </resources>
                             </exception>
                         </exceptions>
+                        <ignoredResourcePatterns combine.children="append">
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply pom.xml:2074]: It doesn't propagate so I'll have to duplicate the logic in downstream dependents. So probably best to leave here? I added a comment.

```
[ERROR] Failed to execute goal org.basepom.maven:duplicate-finder-maven-plugin:1.5.0:check (default) on project trino-redis: Found duplicate classes/resources! -> [Help 1]
[ERROR] Failed to execute goal org.basepom.maven:duplicate-finder-maven-plugin:1.5.0:check (default) on project trino-kinesis: Found duplicate classes/resources! -> [Help 1]
[ERROR] Failed to execute goal org.basepom.maven:duplicate-finder-maven-plugin:1.5.0:check (default) on project trino-kafka: Found duplicate classes/resources! -> [Help 1]
```

[the remark this answers is not in the document]
Comment (nevillelyh) [reply pom.xml]: Actually seems we're doing the same with AWS artifacts. Better to keep the rules localized.
`<!-- com.amazonaws:aws-java-sdk-core and software.amazon.awssdk:sdk-core MIME type file duplicate-->`

**Comment by nevillelyh** [comment]
@Praveen2112 localized `duplicate-finder-maven-plugin`, fixed a few more static imports, going to leave `wire-schema` version as is for this change. Should be good now.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java (hunk 8 above)
Comment (nevillelyh) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufRowDecoder.java]: Pretty sure I did but must've lost it during rebase. Just fixed.

**Comment by Praveen2112** [comment]
Thanks for working on this.

**Comment by ignitz** [comment]
Holy shit guys. That's awesome. 🎉

**Comment by Giovarco** [comment]
Hello! I'm on version 403 and can't handle protobuf messages in any way (as far as I understood) and this protobuf support is nice to have! I'd like to test this. Have you documented anywhere how to use this?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            return new ArrayType(baseType);
+        }
+        return baseType;
+    }
+
+    private Type getTypeForMessage(FieldDescriptor fieldDescriptor)
+    {
+        Descriptor descriptor = fieldDescriptor.getMessageType();
+        if (fieldDescriptor.getMessageType().getFullName().equals(TIMESTAMP_TYPE_NAME)) {
+            return createTimestampType(6);
```
Comment (findepi) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java:102]: Per https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp

> A Timestamp represents a point in time independent of any time zone or calendar, represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time

I.e. it's an instant.
We don't have an instant type (https://github.com/trinodb/trino/issues/2273), so we represent point-in-time data as "timestamp with time zone".

Would you consider changing the mapping to reveal the point-in-time semantics of the data?

Comment (findepi) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java:102]: Also, the type is nano-second precision. Why do we map it to microsecond precision?

cc @mx123

Comment (nevillelyh) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java:102]: These are direct ports from SEP, maybe @Praveen2112 remembers why?
Changing the semantic or precision might break back-compat, we might have to support both local/instant x all precisions combo?

Comment (findepi) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java:102]: > Changing the semantic or precision might break back-compat,

Of course. Every change, including bug fixes, can break something.

>  we might have to support both local/instant x all precisions combo?

not sure why we would want that.

**Pull request merged**